numpy/core/src/multiarray/scalarapi.c

```diff
@@ -87,83 +87,12 @@ scalar_value(PyObject *scalar, PyArray_Descr *descr)
     }
 
     /*
-     * Must be a user-defined type --- check to see which
-     * scalar it inherits from.
+     * Must be a user defined type with an associated (registered) dtype.
+     * Thus, it cannot be flexible (user dtypes cannot be), so we can (and
+     * pretty much have no choice but to) assume the below logic always works.
+     * I.e. this assumes that the logic would also work for most of our types.
      */
 
-#define _CHK(cls) PyObject_IsInstance(scalar, \
-            (PyObject *)&Py##cls##ArrType_Type)
-#define _IFCASE(cls) if (_CHK(cls)) return &PyArrayScalar_VAL(scalar, cls)
-
-    if (_CHK(Number)) {
-        if (_CHK(Integer)) {
-            if (_CHK(SignedInteger)) {
-                _IFCASE(Byte);
-                _IFCASE(Short);
-                _IFCASE(Int);
-                _IFCASE(Long);
-                _IFCASE(LongLong);
-                _IFCASE(Timedelta);
-            }
-            else {
-                /* Unsigned Integer */
-                _IFCASE(UByte);
-                _IFCASE(UShort);
-                _IFCASE(UInt);
-                _IFCASE(ULong);
-                _IFCASE(ULongLong);
-            }
-        }
-        else {
-            /* Inexact */
-            if (_CHK(Floating)) {
-                _IFCASE(Half);
-                _IFCASE(Float);
-                _IFCASE(Double);
-                _IFCASE(LongDouble);
-            }
-            else {
-                /*ComplexFloating */
-                _IFCASE(CFloat);
-                _IFCASE(CDouble);
-                _IFCASE(CLongDouble);
-            }
-        }
-    }
-    else if (_CHK(Bool)) {
-        return &PyArrayScalar_VAL(scalar, Bool);
-    }
-    else if (_CHK(Datetime)) {
-        return &PyArrayScalar_VAL(scalar, Datetime);
-    }
-    else if (_CHK(Flexible)) {
-        if (_CHK(String)) {
-            return (void *)PyString_AS_STRING(scalar);
-        }
-        if (_CHK(Unicode)) {
```

Comment:
Is it worth throwing an exception for these cases if we no longer support them?

Reply:
Well, this function can't write an error easily. As mentioned in the comment, I do not think it is reachable. Flexible user dtypes are unsupported. So you would need something like a fixed-width string dtype which somehow still subclasses our flexible strings.  On top of that you would need to not use getitem/setitem (which is encouraged) or have some external user use low-level C-API...

But I could check for inheriting from `np.flexible` and return NULL just to provoke a certainly hard crash...